### PR TITLE
perf: compute-efficiency optimizations (Qwen CFG merge, FSDP/H2D prefetch, comm fusion)

### PR DESCRIPTION
**Pull request by Jayce-Ping**
## Summary

Compute-efficiency optimizations for the RL trainers + models, plus a follow-up
that extends the non-blocking H2D offload path to **every** trainer and cleans up
the shared utilities. All changes are correctness-preserving (numerically
equivalent) and individually validated.

### Performance changes
- **Qwen-Image CFG single batched forward** — merge the cond/uncond classifier-free
  guidance passes into one batched transformer forward (rollout + training log-prob),
  recovering GPU occupancy at small per-device batch and halving kernel launches /
  `cache_context` overhead. Same FLOPs. Applies to `qwen_image` and
  `qwen_image_edit_plus`.
- **Non-blocking H2D sample prefetch under CPU offload** — a shared
  `_iter_prefetched_batches` / `_iter_prefetched_sample_batches` double-buffers the
  next micro-batch's host→device copy on a dedicated copy stream so it overlaps the
  current batch's compute. Adopted by GRPO/GRPO-Guard/NFT/AWM/DPPO/DPO/OPD; DGPO/CRD
  get the H2D-correctness fix their offload path was missing.
- **FSDP1 `forward_prefetch`** enabled in the FSDP1 accelerate configs (overlaps the
  next layer's param all-gather with current compute; grows with depth / world size
  / interconnect latency). FSDP2 prefetches implicitly.
- **Comm hot-path micro-opt** — fuse the min/max all-reduces into a single MIN
  reduce (`max(h) == -min(-h)`) and drop a redundant `empty_cache` in the trajectory
  gather.

### Models / correctness
- Drop the deprecated Qwen `txt_seq_lens` (diffusers ≥ 0.38 derives it from the
  attention mask); bump `diffusers>=0.38.0` in `pyproject.toml`.
- Per-adapter `ddp_find_unused_parameters` (Wan/Bagel opt in; default False lets DDP
  overlap grad all-reduce and fail fast otherwise).
- `BaseSample.to()` now moves tensors nested inside `extra_kwargs` (and nested
  dict/list) across offload/prefetch.

### Cleanup
- Consolidate three hand-rolled nested-tensor walks into
  `utils.base.map_tensor_leaves` / `visit_tensor_leaves`; share Qwen `_pad_seq_dim`;
  simplify `_pad_batch_prompt`. Tighten docstrings.

## Benchmark (branch vs `main`, 8× H20, same env, seed=42)

Baseline is unmodified `main` run via a git worktree (`PYTHONPATH`), so only the
code differs.

| Optimization | Workload | Metric | main | branch | Speedup |
|---|---|---|---|---|---|
| Qwen CFG merged forward | Qwen-Image LoRA, 8 GPU, 16 steps | compute phase (1 optimize + 1 rollout) | 106 s | 77 s | **~27%** |
| Qwen CFG (same run) | " | total wall-clock | 351.3 s | 298.6 s | ~15% |
| FSDP1 forward_prefetch | FLUX.2-klein-4B full FSDP1, 8 GPU | total wall-clock | 160.6 s | 155.4 s | ~3% |
| merged min/max all-reduce | 8-GPU microbench | `global_tensor_stats_batch` ms/call | 1.513 | 1.480 | ~2% |
| drop gather empty_cache | 8-GPU microbench | `all_gather_tensor_list` ms/call | 1.304 | 1.254 | ~4% |
| H2D prefetch (offload) | Wan2.1-T2V-14B, 33f@480p, 8 GPU | total wall-clock | 334.5 s | 337.7 s | ~0% (compute-bound) |

Headline win is the Qwen CFG merge (~27% on the compute phase, largest at the
small per-device batch typical of RL). The heavy-video Wan case is **compute-bound**
(GPUs already ~98% busy), so the H2D prefetch correctly neither helps nor hurts
there; its wall-clock benefit appears in H2D-bound regimes. C/D are small but
broadly applicable and scale with model size / world size / interconnect latency.

## Correctness / validation

- **H2D offload parity** (offload on vs off, same seed, step-0 byte-identical):
  DPPO `0.1982`, DPO `0.6931`, DGPO `-0.0085`, CRD `0.0091`, OPD `kl_div=0.0004`.
  DGPO/CRD offload was previously broken (CPU batch never moved to device) — now fixed.
- **Qwen CFG merge**: transformer-level parity (bf16) on valid tokens for T2I +
  edit-plus; 32-GPU DDP e2e keeps `ratio ≈ 1.0` (train/inference consistency).
- Cross-task: Qwen-Image 20B (t2i), Wan2.1-14B (t2v), FLUX.2-klein (t2i).
- `pytest tests/` — 19 passed (min/max fusion, Qwen CFG parity, sample offload move).

## Test plan

- [x] Per-optimization numerical parity vs `main` (step-0 / transformer-level)
- [x] H2D offload on/off byte-identical for all 6 adopting trainers
- [x] Qwen CFG `ratio ≈ 1.0` on 32-GPU DDP
- [x] `pytest tests/` green
- [ ] Reviewer sanity check on a representative model in your environment

### CODE REVIEW

File: src/flow_factory/trainers/loader.py [hunk 1]
```diff
@@ -58,6 +58,16 @@ def load_trainer(config: Arguments) -> BaseTrainer:
...
         trainer = load_trainer(config)
     """
+    # Resolve DDP find_unused_parameters from the adapter class (opt-in per
+    # model). Resolving via the registry imports only the class (no
+    # instantiation). This kwarg only affects the DDP backend; FSDP/DeepSpeed
+    # ignore it. Default False lets DDP use static buckets and overlap gradient
+    # all-reduce with backward; adapters that leave trainable params ungraded in
+    # some iterations (e.g. Qwen-Image) opt in via ddp_find_unused_parameters.
+    adapter_cls = get_model_adapter_class(config.model_args.model_type)
+    find_unused = getattr(adapter_cls, "ddp_find_unused_parameters", False)
```
Comment (Jayce-Ping) [inline src/flow_factory/trainers/loader.py:68]: **[Medium]** Default is now `False` (per-adapter opt-in). Opted in: Qwen, Wan, Bagel; validated under DDP: Qwen-LoRA / FLUX.2-klein / SD3.5. Adapters not opted-in and not re-tested (flux1, flux1_kontext, sdxl, sd3, ...) that relied on the old global `True` will fail-fast with "didn't receive grad" if they leave a trainable param ungraded in some iteration. Suggest a quick 2-GPU DDP smoke per remaining adapter, or opt them in.

File: src/flow_factory/models/abc.py [hunk 2]
```diff
@@ -137,6 +137,15 @@ class BaseAdapter(ABC):
...
     LATENT_AXES: ClassVar[Optional[LatentAxes]] = None
 
+    # DDP-only knob (ignored under FSDP/DeepSpeed). Set True only for adapters
+    # whose training step can leave some trainable parameters without a gradient
+    # in a given iteration (e.g. Qwen-Image: guidance=None leaves the guidance
+    # embedder unused). The default False lets DDP use static buckets and overlap
+    # gradient all-reduce with backward; if an adapter actually needs True but
+    # leaves it False, DDP fails fast at the first backward with an explicit
+    # "didn't receive grad" error rather than silently producing wrong results.
+    ddp_find_unused_parameters: ClassVar[bool] = False
```
Comment (Jayce-Ping) [inline src/flow_factory/models/abc.py:147]: This flips the effective default from the old global `True` (loader.py) to `False`. Good for DDP bucket overlap + fail-fast, but please confirm every DDP-trainable adapter either has no conditionally-unused trainable params or opts in.

File: src/flow_factory/samples/samples.py [hunk 3]
```diff
@@ -266,21 +267,49 @@ def tensor_to_repr(v):
...
+                setattr(self, field.name, _move(value))
             elif depth == 1 and is_tensor_list(value):
                 setattr(
                     self,
                     field.name,
-                    [t.to(device) if isinstance(t, torch.Tensor) else t for t in value]
+                    [_move(t) if isinstance(t, torch.Tensor) else t for t in value]
                 )
-            
+            elif depth == 1 and isinstance(value, dict):
```
Comment (Jayce-Ping) [inline src/flow_factory/samples/samples.py:312]: **[Low]** This moves tensors in *every* dict field at depth==1, not just `extra_kwargs`. Fine today (only extra_kwargs holds tensors) but broader than the comment implies; consider scoping to `extra_kwargs` or updating the comment.

**Review by Jayce-Ping — COMMENTED** [review]
## Code review — compute-efficiency PR

Reviewed the full branch diff (15 commits). **No hard bugs found** — an independent Bugbot pass also reported none, and every behavioral change has step-0 byte-identical / `ratio≈1.0` validation. The findings below are behavioral-regression **risks** and completeness gaps, ordered by severity. Each has a concrete fix; I'll apply them after your review.

### Findings

**1. [Medium] DDP `find_unused_parameters`: untested adapters may regress**
`loader.py` replaced the global `DistributedDataParallelKwargs(find_unused_parameters=True)` with a per-adapter opt-in that defaults to `False`. Opted in: Qwen-Image (+Edit-Plus), Wan (t2v/i2v/v2v), Bagel. Validated under DDP: Qwen-LoRA, FLUX.2-klein, SD3.5 (OPD). **Not opted-in and not tested**: `flux1`, `flux1_kontext`, `sdxl`, `sd3`, and any other adapter that was implicitly relying on the old global `True`. If such an adapter leaves any *trainable* param ungraded in an iteration, DDP now raises "didn't receive grad". It is fail-fast (loud, not silent) and a one-line opt-in fixes it, but it is a regression surface for adapters nobody re-ran.
- *Fix:* 2-GPU DDP smoke per remaining adapter (or audit each forward for conditionally-unused trainable params); opt-in those that need it; optionally catch the DDP RuntimeError and re-raise with a hint naming `ddp_find_unused_parameters`.

**2. [Low-Med] No committed regression tests (`tests/` is gitignored)**
The parity/unit tests (min/max fusion, Qwen CFG parity, sample offload move) and the new helpers are validated locally only; `tests/` has 0 tracked files, so the PR's "pytest 19 passed" is not CI-reproducible. The pure functions (`map_tensor_leaves`, `visit_tensor_leaves`, dist min/max fusion, `BaseSample.to(extra_kwargs)`, `_pad_seq_dim`) are trivially unit-testable.
- *Fix:* move the CPU-testable units into a tracked test path so the parity guarantees are enforced in CI.

**3. [Low] Dropping `empty_cache` trades memory headroom for speed** — `utils/dist.py:all_gather_tensor_list`
Removing `torch.cuda.empty_cache()` removes a device sync (the win) but also the cache-pool drop; a very memory-tight, fragmented run could now OOM where it previously survived.
- *Fix:* keep as-is; note in the changelog that tight runs may want `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`.

**4. [Low] `BaseSample.to()` now moves tensors in *all* dict fields, not just `extra_kwargs`** — `samples.py:308`
The new dict branch runs `map_tensor_leaves` on every dict field at `depth==1`. Today only `extra_kwargs` holds tensors, so it's fine, but it's broader than the comment implies.
- *Fix:* scope to `extra_kwargs`, or update the comment to "all dict fields".

**5. [Low] `pin_memory` path double-copies (pageable → pinned)** — `samples.py:_move`
`t.to('cpu').pin_memory()` does a D2H to pageable then a host copy into pinned memory. Correct, minor extra host copy per offloaded tensor.
- *Fix:* optional; only if D2H offload time shows up in a profile.

**6. [Nit] `_pad_batch_prompt` `max_pos_len` can be 0 for an all-zero mask**
Degenerate/empty-prompt mask → `int(mask.sum(...).max()) == 0` → padding to length 0. Practically impossible (prompts are non-empty), but a `max(1, ...)` guard would be defensive.

### Benchmark
See PR description for the full branch-vs-`main` table (headline: Qwen CFG merge ~27% on the compute phase). Fresh Wan2.1-T2V-14B (33f@480p) A/B re-confirms the heavy-video case is compute-bound (~parity; GPUs ~100%): **branch ~321 s**.

### Bugbot
No confirmed bugs.

**Comment by Jayce-Ping** [comment]
### Fresh benchmark re-confirmation (latest branch vs `main`)

**Wan2.1-T2V-14B, 33 frames @ 480p, offload on, 8x H20** (heavy model + long video):

| | total wall-clock |
|---|---|
| branch (latest) | **320.8 s** |
| `main` (baseline) | 324.7 s |

~parity (branch ~1.2% faster, within noise) -- the heavy-video workload is **compute-bound**, GPUs at ~100% on both, so the H2D prefetch correctly neither helps nor hurts here. No regression on the heavy target. (The headline speedup remains the Qwen CFG merge, ~27% on the compute phase; see the table in the PR description.)

**Bugbot** independent review: no confirmed bugs.

**Comment by Jayce-Ping** [comment]
### Review follow-up — fixes pushed (`e0980f6`)

- **[Medium] DDP `find_unused_parameters`** — audited all 6 un-opted-in adapters (`z_image`, `ltx2_t2av`, `ltx2_i2av`, `flux2`, `flux1`, `flux1_kontext`). None have conditionally-unused trainable params: Flux2/Flux1/Kontext always pass a non-`None` guidance tensor (unlike Qwen/Flux2-Klein `guidance=None`); none route between two trainable transformers per step (unlike Wan); none drop expert groups (unlike Bagel); LTX2's primary forward always runs both video+audio with cross-modal attention. **Default `False` is correct for all six; no opt-in needed.** (Edge case noted: a SigLIP-enabled Z-Image *Omni* checkpoint full-fine-tuned via the T2I path, outside normal configs.)
- **[Low-Med] Tests** — `tests/` is intentionally gitignored ("kept out of the public repo"); kept local per that policy. Validation stays documented in the PR.
- **[Low] `BaseSample.to` dict-move** — comment clarified (applies to any dict field; in practice only `extra_kwargs`).
- **[Low] `empty_cache`** — behavior kept; doc note dropped since the compute-efficiency prose was trimmed.
- **[Low] `pin_memory` double-copy** — fixed: copy straight into a pinned buffer (one D2H).
- **[Nit] `_pad_batch_prompt`** — added `max(1, ...)` guard.

Also per maintainer request: trimmed the redundant compute-efficiency knob prose (`workflow.md` + multinode README; inline config comments kept), and **lowered the diffusers floor 0.38.0 -> 0.37.0** (0.37.0 already derives the text seq len from the mask and ships `QwenImageEditPlusPipeline`; 0.36.0 does not).

`pytest tests/` 19 passed; imports clean.
